akka-actor/src/main/scala/akka/dispatch/affinity/AffinityPool.scala

```diff
@@ -0,0 +1,401 @@
+import java.util
+import java.util.Collections
+import java.util.concurrent.TimeUnit.MICROSECONDS
+import java.util.concurrent._
+import java.util.concurrent.atomic.{ AtomicInteger, AtomicReference }
+import java.util.concurrent.locks.{ Lock, LockSupport, ReentrantLock }
+
+import akka.dispatch._
+import akka.dispatch.affinity.AffinityPool._
+import akka.util.Helpers.Requiring
+import com.typesafe.config.Config
+
+import scala.annotation.tailrec
+import java.lang.Integer.reverseBytes
+
+import akka.util.ImmutableIntMap
+import akka.util.OptionVal
+
+import scala.collection.mutable
+import scala.util.control.NonFatal
+
+private[akka] class AffinityPool(
+  parallelism:               Int,
+  affinityGroupSize:         Int,
+  tf:                        ThreadFactory,
+  idleCpuLevel:              Int,
+  fairDistributionThreshold: Int,
+  rejectionHandler:          RejectionHandler)
+  extends AbstractExecutorService {
+
+  if (parallelism <= 0)
+    throw new IllegalArgumentException("Size of pool cannot be less or equal to 0")
+
+  // Held while starting/shutting down workers/pool in order to make
+  // the operations linear and enforce atomicity. An example of that would be
+  // adding a worker. We want the creation of the worker, addition
+  // to the set and starting to worker to be an atomic action. Using
+  // a concurrent set would not give us that
+  private val bookKeepingLock = new ReentrantLock()
+
+  // condition used for awaiting termination
+  private val terminationCondition = bookKeepingLock.newCondition()
+
+  // indicates the current state of the pool
+  @volatile final private var poolState: PoolState = Running
+
+  private final val workQueues = Array.fill(parallelism)(new BoundedTaskQueue(affinityGroupSize))
+  private final val workers = mutable.Set[ThreadPoolWorker]()
+
+  // a counter that gets incremented every time a task is queued
+  private val executionCounter: AtomicInteger = new AtomicInteger(0)
+  // maps a runnable to an index of a worker queue
+  private val runnableToWorkerQueueIndex = new AtomicReference(ImmutableIntMap.empty)
+
+  private def getQueueForRunnable(command: Runnable) = {
+
+    val runnableHash = command.hashCode()
+
+    def sbhash(i: Int) = reverseBytes(i * 0x9e3775cd) * 0x9e3775cd
+
+    def getNext = executionCounter.incrementAndGet() % parallelism
+
+    def updateIfAbsentAndGetQueueIndex(
+      workerQueueIndex: AtomicReference[ImmutableIntMap],
+      runnableHash:     Int, queueIndex: ⇒ Int): Int = {
+      @tailrec
+      def updateIndex(): Unit = {
+        val prev = workerQueueIndex.get()
+        if (!runnableToWorkerQueueIndex.compareAndSet(prev, prev.updateIfAbsent(runnableHash, queueIndex))) {
+          updateIndex()
+        }
+      }
+      updateIndex()
+      workerQueueIndex.get().get(runnableHash) // can safely call get..
+    }
+
+    val workQueueIndex =
+      if (runnableToWorkerQueueIndex.get().size > fairDistributionThreshold)
```

Comment:
That unfortunately would disable it forever. Better would be to put smth like: 

```
    val workQueueIndex =
      if (fairDistributionThreshold == 0 || runnableToWorkerQueueIndex.get().size > fairDistributionThreshold)
        Math.abs(sbhash(runnableHash)) % parallelism
      else
        updateIfAbsentAndGetQueueIndex(runnableToWorkerQueueIndex, runnableHash, getNext)
```

I have added in the docs that a value of 0 means no map based approach at all.